// mx-react-components/dist/components/d3/AxisGroup.js
'use strict';

var React = require('react');
var d3 = require('d3');

var ChartUtils = require('../../utils/Chart');

var AxisGroup = React.createClass({
  displayName: 'AxisGroup',

  propTypes: {
    axis: React.PropTypes.string.isRequired,
    axisFormatFunction: React.PropTypes.func.isRequired,
    data: React.PropTypes.array.isRequired,
    orientation: React.PropTypes.string.isRequired,
    scaleFunction: React.PropTypes.func.isRequired,
    translation: React.PropTypes.string
  },

  getDefaultProps: function getDefaultProps() {
    return {
      translation: 'translate(0,0)'
    };
  },
  componentDidMount: function componentDidMount() {
    this._renderAxis();
  },
  componentDidUpdate: function componentDidUpdate() {
    this._renderAxis();
  },
  _renderAxis: function _renderAxis() {
    var tickValues = ChartUtils.getAxisTickValues(this.props.data, this.props.axis);
    var axisFunction = d3.svg.axis().scale(this.props.scaleFunction()).orient(this.props.orientation).tickFormat(this.props.axisFormatFunction).ticks(tickValues.length).tickValues(tickValues);

    d3.select(this.refs[this.props.axis + 'Axis']).call(axisFunction);
  },
  render: function render() {
    return React.createElement('g', {
      className: this.props.axis + '-axis',
      ref: this.props.axis + 'Axis',
      transform: this.props.translation
    });
  }
});

module.exports = AxisGroup;
