// mx-react-components/dist/components/d3/GridLinesGroup.js
'use strict';

var React = require('react');

var d3 = require('d3');

var ChartUtils = require('../../utils/Chart');

var GridLinesGroup = React.createClass({
  displayName: 'GridLinesGroup',

  propTypes: {
    axis: React.PropTypes.string.isRequired,
    data: React.PropTypes.array.isRequired,
    orientation: React.PropTypes.string,
    scaleFunction: React.PropTypes.func.isRequired,
    tickSize: React.PropTypes.number.isRequired,
    translation: React.PropTypes.string
  },

  getDefaultProps: function getDefaultProps() {
    return {
      orientation: 'left',
      translation: 'translate(0,0)'
    };
  },
  componentDidMount: function componentDidMount() {
    this._renderGridLines();
  },
  componentDidUpdate: function componentDidUpdate() {
    this._renderGridLines();
  },
  _renderGridLines: function _renderGridLines() {
    var tickValues = ChartUtils.getAxisTickValues(this.props.data, this.props.axis);

    var gridLinesFunction = d3.svg.axis().scale(this.props.scaleFunction()).orient(this.props.orientation).tickSize(this.props.tickSize, 0, 0).tickFormat('').ticks(tickValues.length).tickValues(tickValues);

    d3.select(this.refs[this.props.axis + 'GridLines']).call(gridLinesFunction);
  },
  render: function render() {
    return React.createElement('g', {
      className: this.props.axis + '-grid-line',
      ref: this.props.axis + 'GridLines',
      transform: this.props.translation
    });
  }
});

module.exports = GridLinesGroup;
