// mx-react-components/dist/components/d3/LineGroup.js
'use strict';

var React = require('react');

var d3 = require('d3');

var StyleConstants = require('../../constants/Style');

var LineGroup = React.createClass({
  displayName: 'LineGroup',

  propTypes: {
    adjustedHeight: React.PropTypes.number.isRequired,
    dashLine: React.PropTypes.bool,
    data: React.PropTypes.array.isRequired,
    lineColor: React.PropTypes.string,
    shouldAnimate: React.PropTypes.bool,
    strokeWidth: React.PropTypes.number,
    translation: React.PropTypes.string,
    xScaleValueFunction: React.PropTypes.func.isRequired,
    yScaleValueFunction: React.PropTypes.func.isRequired
  },

  getDefaultProps: function getDefaultProps() {
    return {
      dashLine: false,
      lineColor: StyleConstants.Colors.CHARCOAL,
      shouldAnimate: true,
      strokeWidth: 2,
      translation: 'translate(0,0)'
    };
  },
  componentWillMount: function componentWillMount() {
    var _this = this;

    var flatLine = d3.svg.line().x(function (d) {
      return _this.props.xScaleValueFunction(d.x);
    }).y(function () {
      return _this.props.adjustedHeight;
    });

    var line = d3.svg.line().x(function (d) {
      return _this.props.xScaleValueFunction(d.x);
    }).y(function (d) {
      return _this.props.yScaleValueFunction(d.y);
    });

    this.setState({
      flatLine: flatLine,
      line: line
    });
  },
  componentDidMount: function componentDidMount() {
    this._animateLine();
  },
  componentDidUpdate: function componentDidUpdate() {
    this._animateLine();
  },
  _animateLine: function _animateLine() {
    if (this.props.shouldAnimate) {
      d3.select(this.refs.chartLine).transition().attr('d', this.state.line(this.props.data));
    }
  },
  render: function render() {
    var _props = this.props;
    var data = _props.data;
    var dashLine = _props.dashLine;
    var lineColor = _props.lineColor;
    var shouldAnimate = _props.shouldAnimate;
    var strokeWidth = _props.strokeWidth;
    var translation = _props.translation;


    return React.createElement(
      'g',
      { className: 'chart-line-group', ref: 'chartLineGroup', transform: translation },
      React.createElement('path', {
        d: shouldAnimate ? this.state.flatLine(data) : this.state.line(data),
        fill: 'none',
        ref: 'chartLine',
        stroke: lineColor,
        strokeDasharray: dashLine ? '4,4' : 'none',
        strokeWidth: strokeWidth
      })
    );
  }
});

module.exports = LineGroup;
