// mx-react-components/dist/components/d3/TimeXAxisGroup.js
'use strict';

var React = require('react');

var d3 = require('d3');
var moment = require('moment');

var TimeXAxisGroup = React.createClass({
  displayName: 'TimeXAxisGroup',

  propTypes: {
    numberOfTicks: React.PropTypes.number,
    tickSize: React.PropTypes.number,
    timeAxisFormat: React.PropTypes.string.isRequired,
    translation: React.PropTypes.string,
    xScaleFunction: React.PropTypes.func.isRequired
  },

  getDefaultProps: function getDefaultProps() {
    return {
      numberOfTicks: 10,
      tickSize: 6,
      translation: 'translate(0,0)'
    };
  },
  componentDidMount: function componentDidMount() {
    this._renderAxis();
  },
  componentDidUpdate: function componentDidUpdate() {
    this._renderAxis();
  },
  _renderAxis: function _renderAxis() {
    var _this = this;

    var timeAxisFunction = d3.svg.axis().scale(this.props.xScaleFunction()).tickFormat(function (d) {
      return moment.unix(d).format(_this.props.timeAxisFormat);
    }).tickSize(this.props.tickSize, this.props.tickSize).ticks(this.props.numberOfTicks);

    d3.select(this.refs.timeAxis).call(timeAxisFunction);
  },
  render: function render() {
    return React.createElement('g', {
      className: 'time-axis',
      ref: 'timeAxis',
      transform: this.props.translation
    });
  }
});

module.exports = TimeXAxisGroup;
