// mx-react-components/dist/components/SimpleInput.js
'use strict';

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var React = require('react');
var _merge = require('lodash/merge');

var Icon = require('./Icon');
var StyleConstants = require('../constants/Style');

var Input = React.createClass({
  displayName: 'Input',

  propTypes: {
    baseColor: React.PropTypes.string,
    focusOnLoad: React.PropTypes.bool,
    icon: React.PropTypes.string,
    placeholder: React.PropTypes.string,
    style: React.PropTypes.oneOfType([React.PropTypes.array, React.PropTypes.object]),
    styles: React.PropTypes.object,
    type: React.PropTypes.string,
    valid: React.PropTypes.bool
  },

  getDefaultProps: function getDefaultProps() {
    return {
      baseColor: StyleConstants.Colors.PRIMARY,
      focusOnLoad: false,
      type: 'text',
      valid: true
    };
  },
  getInitialState: function getInitialState() {
    return {
      focus: false
    };
  },
  componentDidMount: function componentDidMount() {
    if (this.props.style) {
      console.warn('The style prop is deprecated and will be removed in a future release. Please used styles.');
    }

    if (this.props.focusOnLoad) {
      this.refs.input.focus();
    }
  },
  _onFocus: function _onFocus() {
    this.refs.input.focus();

    this.setState({
      focus: true
    });
  },
  _onBlur: function _onBlur() {
    this.refs.input.blur();

    this.setState({
      focus: false
    });
  },
  render: function render() {
    var styles = this.styles();

    return React.createElement(
      'div',
      {
        onBlur: this._onBlur,
        onFocus: this._onFocus,
        style: _extends({}, styles.wrapper, this.state.focus ? styles.activeWrapper : null),
        tabIndex: 0
      },
      this.props.icon ? React.createElement(Icon, { size: 20, style: styles.icon, type: this.props.icon }) : null,
      React.createElement('input', _extends({}, this.props, {
        ref: 'input',
        style: styles.input,
        type: this.props.type
      }))
    );
  },
  styles: function styles() {
    return _merge({}, {
      wrapper: _extends({}, {
        padding: StyleConstants.Spacing.SMALL,
        boxSizing: 'border-box',
        backgroundColor: StyleConstants.Colors.WHITE,
        border: this.props.valid ? '1px solid ' + StyleConstants.Colors.FOG : '1px solid ' + StyleConstants.Colors.STRAWBERRY,
        borderRadius: 3,
        display: 'flex',
        alignItems: 'center',
        width: '100%',
        outline: 'none',
        boxShadow: 'none'
      }, this.props.style),
      activeWrapper: {
        border: '1px solid ' + this.props.baseColor
      },
      icon: {
        paddingRight: StyleConstants.Spacing.XSMALL,
        fill: this.props.baseColor
      },
      input: {
        flex: '1 0 0%',
        backgroundColor: StyleConstants.Colors.WHITE,
        border: 'none',
        outline: 'none',
        boxShadow: 'none'
      }
    }, this.props.styles);
  }
});

module.exports = Input;
